Complexity_classes_def/IMP-_Reductions.thy

```diff
@@ -0,0 +1,166 @@
+  hence "comp (f;;g) x z" using comp_comp f_part(1)_g_part(1) by auto
+  moreover have "x \<in>D \<longleftrightarrow> 0 = z" using f_part(2) g_part(2) by blast
+  ultimately show "\<exists>r. Abstractions.comp (f;; g) x r \<and> (x \<in> D) = (r = 0)" by blast
+qed
+
+lemma reduction_verification_correct:
+  assumes "is_reduction f D D'" "is_verif g D'"
+  shows "is_verif (f;;g) D"         
+  apply(simp add:is_verif_def)
+proof
+  show "\<forall>x z. \<exists>r. verif (f;; g) x z r"
+  proof fix x 
+    show "\<forall>z. \<exists>r. verif (f;; g) x z r" 
+    proof fix z
+      from assms(1) obtain y where  "comp f x y " using is_reduction_def by meson
+      moreover from assms(2) obtain r where  "verif g y z r" using is_verif_def by meson
+      moreover have "cons_certif f" using assms(1) is_reduction_def  by simp
+      ultimately show "\<exists>r. verif (f;;g) x z r" using comp_verif by blast
+      qed
+    qed
+next
+  show "\<forall>x. (x \<in> D) = (\<exists>z. verif (f;; g) x z 0)"
+  proof
+    fix x 
+    from assms(1) obtain y where y_def:  "comp f x y " " (x\<in>D) = (y\<in>D')" 
+      using is_reduction_def by meson
+    moreover from assms(2) have "(y\<in>D')  = (\<exists>z. verif g y z 0)" 
+      using is_verif_def by simp 
+    ultimately have "(x \<in> D) = (\<exists>z. verif g y z 0) " by simp
+    moreover have "cons_certif f" using assms(1) is_reduction_def  by simp
+    ultimately show "(x \<in> D) = (\<exists>z. verif (f;; g) x z 0)" using y_def(1) comp_verif
+      by (metis (full_types) assms(2) is_verif_def verif_det) 
+  qed
+qed
+
+
+lemma is_reduction_compose:
+  assumes "is_reduction f D D'" "is_reduction g D' D''"
+  shows "is_reduction (f;;g) D D''"
+  apply(auto simp add: is_reduction_def)
+  using assms apply(simp add: cons_certif_def is_reduction_def) apply fastforce
+proof -
+  fix x
+   obtain y where y_def: "comp f x y" "(x\<in>D) =(y\<in>D')"
+     using assms(1) by (auto simp add:is_reduction_def)
+   obtain z where z_def: "comp g y z" "(y\<in>D') =(z\<in>D'')"
+     using assms(2) by (auto simp add:is_reduction_def)    
+   have  "comp (f;; g) x z  \<and> (x \<in> D) = (z \<in> D'')"
+     using y_def z_def comp_comp by simp
+   thus  "\<exists>r. comp (f;; g) x r \<and> (x \<in> D) = (r \<in> D'')" by blast
+ qed
+
+subsection \<open>Polynomial reductions\<close>
+paragraph \<open>Definition\<close>
+text \<open>
+A polyreduction is a reduction that is also polynomially bounded in time (in most literature)
+
+A necessary condition also for polyreduction is to be polynomially result-bounded
+(otherwise polyreduction would not be transitive)
+Yet, this condition is generally omitted in most literature, as TM is the commonly used model
+of computation, where the limitation in time implies a limitation in space. 
+As it is not the case in our model, we choose to add explicitly this condition.
+
+It is yet worth the mention, that given that program is a reduction hence
```

Comment:
Can't we also write programs in IMP- that only take input in one register and still produce output that is not polynomially bounded in the running time?

Reply:
if you don't make the restriction that the output is only depending on that single register than yeah. Otherwise that register could contain anything and there's no bound you can make that is independent of the initial state